### DISCUSSION TITLE
https fails, http works what is the reason?

### DISCUSSION THREAD

**Discussion by Otto404** [dotnet/yarp#1827] [Q&A] [answered]
Why can't I use the browser to access an API behind a remote proxy when the connection between the proxy and the API is over https?

or in other words:
What I don't understand: what is the Edge doing wrong, what HttpClient is doing right

Yarp acts as a reverse proxy
It provides access to two web APIs (Food and Drinks)

The client is a Blazor program and using an HttpClient the client can successfully access the api through the proxy.

If I now access the API with the browser (Edge) via the proxy, I get an error (see below). This way only works if I switch the connection between the proxy and the api from https to http. (Food works, drinks not)
I don't understand the reason and would like to know what I'm missing here.

The configuration
``` json
 {
   "Clusters": {
      "food-cluster": {
        "Destinations": {
          "destination1": {
            "Address": "http://localhost:5010"
          }
        }
      },
      "drink-cluster": {
        "Destinations": {
          "destination1": {
            "Address": "https://localhost:5021"
          }
        }
      }
    }
}
```
The Error: 
``` cli
Yarp.ReverseProxy.Forwarder.HttpForwarder: Information: Proxying to https://localhost:5021/drink HTTP/2 RequestVersionOrLower no-streaming
'Gateway.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\6.0.8\System.Diagnostics.StackTrace.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'Gateway.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\6.0.8\System.Reflection.Metadata.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'Gateway.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\6.0.8\System.Collections.Immutable.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
Yarp.ReverseProxy.Forwarder.HttpForwarder: Information: Request: An error was encountered before receiving a response.

System.Net.Http.HttpRequestException: An error occurred while sending the request.
 ---> System.IO.IOException: The request was aborted.
 ---> System.IO.IOException: The response ended prematurely while waiting for the next frame from the server.
   at System.Net.Http.Http2Connection.<ReadFrameAsync>g__ThrowMissingFrame|57_1()
   at System.Net.Http.Http2Connection.ReadFrameAsync(Boolean initialFrame)
   at System.Net.Http.Http2Connection.ProcessIncomingFramesAsync()
   --- End of inner exception stack trace ---
   at System.Net.Http.Http2Connection.ThrowRequestAborted(Exception innerException)
   at System.Net.Http.Http2Connection.Http2Stream.CheckResponseBodyState()
   at System.Net.Http.Http2Connection.Http2Stream.TryEnsureHeaders()
   at System.Net.Http.Http2Connection.Http2Stream.ReadResponseHeadersAsync(CancellationToken cancellationToken)
   at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.DiagnosticsHandler.SendAsyncCore(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at Yarp.ReverseProxy.Forwarder.HttpForwarder.SendAsync(HttpContext context, String destinationPrefix, HttpMessageInvoker httpClient, ForwarderRequestConfig requestConfig, HttpTransformer transformer)
The thread 0x6e9c has exited with code 0 (0x0).
```

**Comment by samsp-msft** [comment, marked as answer]
I fixed your question formatting - its easier to read if you use ``` blocks as you can specify the language for the block and get colorization.

How are the services implemented? What happens if you access them directly with cURL or similar?
It seems that YARP is having an issue with Http/2. H2 typically requires TLS, so changing to http://, will use 1.1 rather H2, so that may be your problem. 
You can configure the http version that YARP will use as part of the destination configuration: https://microsoft.github.io/reverse-proxy/articles/http-client-config.html#httprequest

**Reply by Otto404** [reply]
Thanks for the formatting help.
Both services work well. I can access the Swagger pages and use the services (directly without a proxy).
With Postman, the call succeeds directly and via the proxy. The error only occurs with the browser (Edge).

**Reply by MihaZupan** [reply]
Does the problem go away if you keep the `https` behind the proxy, but change the `HttpRequest.Version` to 1.1 as Sam suggested?
```json
"HttpRequest": {
    "Version": "1.1",
}
```

**Reply by Otto404** [reply]
"HttpRequest": {
    "Version": "1.1",
}
Did the trick.
Sam & Miha, I thank you for your kindness and help

**Reply by Otto404** [reply]
But I have still the original question: 

why(!) is just the browser failing?

**Reply by adityamandaleeka** [reply]
Thanks for checking that 1.1 doesn't have the issue. We should figure out what causes the original issue.

**Reply by adityamandaleeka** [reply]
@Otto404 Can you share logs and any other info from the backend?

Ideally, if you could create a minimal reproducer for this issue and put it on a GitHub repo that would be super helpful for us to investigate.

**Reply by Tratcher** [reply]
Is the browser using HTTPS or HTTP to access YARP? Does it behave any differently depending on which scheme is used by the client?

**Reply by karelz** [reply]
Triage: Not actionable without repro or logs. Marking as Answered. Feel free to ping it when more info is available for us to act on. Thanks!
